### DISCUSSION TITLE
Can we add other regex quantifiers to the rule-based matching?

### DISCUSSION THREAD

**Discussion by randomgambit** [explosion/spaCy#6294] [Help: Coding & Implementations] [answered]
Hello and thanks again for this amazing package.

I was trying to extract `verb-phrases` (VP) using `spacy` and I came across the following issue. Please consider the following example, which only returns `spacy has become`.

```
text = '''
spacy has become a standard.
spacy more and more has generated a lot of buzz.
'''

doc = nlp(text)

pattern = [{"LOWER": "spacy"},
           {'POS': 'VERB', 'OP': '?'},
           {'POS': 'ADV', 'OP': '*'},
           {'POS': 'AUX', 'OP': '*'},
           {'POS': 'VERB', 'OP': '+'}]

matcher.add("VP", None, pattern)

matches = matcher(doc)

for match_id, start, end in matches:
    matched_span = doc[start:end]
    print(matched_span.text)
spacy has become
```

The issue is that as you can see in the second example, `spacy more and more has generated a lot of buzz`, there are 3 tokens between `spacy` and the `VP` `has generated`: "more and more". 

Unfortunately, in a real-life setting, it is hard to know in advance how many tokens there could be between a lemma of interest (here, `spacy`) and the `VP`, so to match correctly my `VP` I would need to create many rules, each having a different number of `{}` in them to indicate the number of "free" tokens between the two.  

For instance:

```
pattern = [{"LOWER": "spacy"},
           {},
           {'POS': 'VERB', 'OP': '?'},
           {'POS': 'ADV', 'OP': '*'},
           {'POS': 'AUX', 'OP': '*'},
           {'POS': 'VERB', 'OP': '+'}]
```

Can we do better here? For instance, you could add a regex quantifier that says `match between n and m times` an empty token.
What do you think?

Thanks!

**Comment by adrianeboyd** [comment, marked as answer]
There's an open issue for quantifiers in matcher patterns: #5603. (We would welcome a PR!)

However, if you want to detect VPs, you'd be better off working with the tags plus the dependency parses, similar to how the noun chunks detection is implemented, e.g.: https://github.com/explosion/spaCy/blob/260c29794a1caa70f8b0702c31fcfecad6bfdadc/spacy/lang/en/syntax_iterators.py

You could also try out the new `DependencyMatcher` that will be in the upcoming spacy v3 release. Install `spacy-nightly` (in a new venv!) to try it out: https://github.com/explosion/spaCy/blob/260c29794a1caa70f8b0702c31fcfecad6bfdadc/spacy/lang/en/syntax_iterators.py

**Comment by randomgambit** [comment]
Thanks @adrianeboyd this is very useful! I will check the dependency parser then. Right now, I was parsing the dependency tree manually using the `token` attributes (children, ancestors, etc).

Just a quick question maybe. I do not understand why you use `left_edge` in https://github.com/explosion/spaCy/blob/260c29794a1caa70f8b0702c31fcfecad6bfdadc/spacy/lang/en/syntax_iterators.py

How is this preventing nested noun chunks? I though a simple algorithm would have been to look for `determiners` and `adjectives` all **children** of a given `noun` in the tree.
```

 for i, word in enumerate(doclike):
        if word.pos not in (NOUN, PROPN, PRON):
            continue
        # Prevent nested chunks from being produced
        if word.left_edge.i <= prev_end:
```

**Comment by randomgambit** [comment]
In  addition, I do not understand why (in Spacy v3) we are using the concept of `left id` and `right id`. I mean, if we have an operator that says `A > B` (A is the head of B), then what is the meaning of left and right? In the dependency tree you can have a subject and a verb in any given order... What do you think?

Thanks again for all the great work!!

**Comment by adrianeboyd** [comment]
`left` and `right` aren't related to the dependency tree or word order, but to the position of the tokens in terms of the semgrex operator.

You could write `A > B` or `B < A` and this would express the same thing in terms of the relation between `A` and `B`, but in the first one `A` is left of the operator and `B` is right of the operator. Because of how the `DependencyMatcher` is implemented, you're restricted to adding the token for the left side first and then adding the operator + right side to extend the pattern.

**Comment by randomgambit** [comment]
thanks for clarifying that. super useful!
